### ISSUE TITLE
keyword "this" e "base" como indicador de membros da classe

### ISSUE THREAD

**Issue by rcktscnc** [Cappta/best-practices#35]
"Chamadas a métodos e variáveis não estáticos em contexto de classe devem sempre utilizar o this. Em caso de herança utilize base"

Gostaria de sugerir a remoção dessa regra. Usar "this" e "base" (exceto em casos onde há necessidade) deixa o código muito poluído. 

Ao meu ver, a única vantagem dessa regra é o (pequeno) contexto proporcionado durante o code review. Mas não acho essa vantagem grande o suficiente para justificar o uso. Em várias ocasiões, durante o review, o trecho de código mostrado no github é tão pequeno que o reviwer vai ficar sem contexto com ou sem o this/base, sendo necessário abrir o arquivo inteiro para ver o que de fato a alteração causa na classe.

**Comment by daniellessio** [comment]
No caso do **this** Concordo que polui. Mas discordo no "muito", nunca pesou para mim, mas pode ser por costume.
Ainda assim, usar a keyword para expressar uma "auto-referência" faz bastante sentido para mim, em métodos que o parâmetro tiver o mesmo nome que o dado da classe, por exemplo. Nesse sentido acho que o ganho de contexto é positivo e não se limita ao review, o código fica mais claro - na minha opinião.

Com relação ao **base**, daí eu já descordo totalmente de você. Se o campo veio da super classe, eu gostaria de saber isso de cara, sem precisar navegar para outros arquivos para saber da onde vem - e depois que eu me deparasse com o primeiro caso disso, faria o _double-check_ de todos os campos para garantir qual a origem destes também.

De qualquer forma, se você está herdando muito, você está fazendo errado, kkkkkkk

**Comment by LouriJr** [comment]
Concordo com o @daniellessio, entendo o ponto de poluição de código, apesar de achar um exagero. Acredito que no trade off os benefícios valem a pena.

**Comment by guMatos** [comment]
Também concordo com o @daniellessio , nunca me incomodei de usar o __this__, existem outras formas de indicar uma auto-referência, mas acredito que o __this__ seja o mais convencional e o "menos feio".

**Comment by vinibeloni** [comment]
Eu concordo tanto com o @rcktscnc  quanto com o @daniellessio , minha posição é o seguinte, na comunidade .net, muita gente usa a propriedade privada com _underline_, e o restante é com ou sem o **_this_** msm. 

sinceramente não vejo tanta vantagem em saber se é ou não da mesma classe ou se é por herança. 

Exemplo, numa _**controller**_, sempre tem métodos **_base_**, mas sempre colocamos **_this_** (pode procurar ai nas APIs da cappta). E não muda absolutamente nada se retirássemos.

Pergunta: Qual desses 3 métodos de retorno(**NotAcceptable**, **InternalServerError**, **NoContext**) não exitem por padrão num projeto .AspNet Core MVC ?

``` csharp
public OrderController : AbstractControllerBase

// com this
  catch (OrderNotFoundException) { return base.NotAcceptable(); }
  catch (TransactionNotFoundException) { return base.NoContent(); }
  catch (Exception) { return base.InternalServerError(); 

// sem this
  catch (OrderNotFoundException) { return NotAcceptable(); }
  catch (TransactionNotFoundException) { return NoContent(); }
  catch (Exception) { return InternalServerError(); 
```


Resposta: só o _**NoContent**_ existe por padrão num projeto .AspNet Core MVC, o resto são herdadas de uma classe base que herda de uma classe chamada _ControllerBase_ do proprio AspNet Core.

Resumo, não mudou em nada ter ou não, a convenção aceitaria numa boa tanto o _this_ pra esses caso, quanto o _base_ e ainda sim vc olhando não saberia ou dificilmente lembraria desses detalhes, e que sinceramente não importaria muito.
 
e mesmo que esses método pertencessem a esta classe e diferenciasse o _this_ do _base_, alguém diria pra criar uma classe base para generalizar o código pra outras _controller_ que precisassem.

outro exemplo (o código original não está dessa forma ainda)
``` csharp
// com this
public Order Handle(GetOrderByIdQuery query)
{
  return this.cacheRepository.Get<Order>(query.OrderId.ToString())
  ??  this.orderRepository.Find(o => o.Id == query.OrderId)
  ??  throw new OrderNotFoundException();
}

// sem this
public Order Handle(GetOrderByIdQuery query)
{
  return cacheRepository.Get<Order>(query.OrderId.ToString())
  ??  orderRepository.Find(o => o.Id == query.OrderId)
  ??  throw new OrderNotFoundException();
}
```

Não mudou em nada, vc sabe q os _Repositories_ vem do construtor porque não está sendo passado como parâmetro pra esse método, e mesmo que fosse já seria estranho demais o cara passar um repositorio como parametro fora de um construtor pra vc lembrar do **_this_**, 

resumindo, eu não usaria em código meu, e também não faria quase diferencia nenhuma porque vc deveria estar muito mais preocupado com cobertura de testes em cima desse método, com o tamanho dele, responsabilidade da classe e etc, um **_this_** mesmo que isso seja automatizado para não passar na convenção, é o menor dos seus problemas, se vc não conseguiu entender algo por conta da ausência de um **_this_** (salvo algumas exceções como nos contrutores por exemplo), o problema está em outro lugar, talvez tamanho do método, nome, parametros, etc. 

Mas enfim, independente se vai usar ou não, temos que ter a ferramente necessária que valide nossas convenções de forma eficaz.

**Comment by daniellessio** [comment]
Com os exemplos do vini, fica claro que o contexto não é prejudicado por falta do _base_ ou do _this_ e concordo que contexto pode (e deve) ser definido muito além do uso dessas keywords.

O ponto principal da convenção para mim é padronização e consistência. Os casos onde tiver necessidade (conforme o comentário do @rcktscnc) podem cair na subjetividade - sobretudo em contextualização através do _this._ Quando a regra é usar isso não ocorre, porque não existe subjetividade - pode esperar que o _this_ estará lá.

Particularmente, acho a falta de _base_ pior do que a falta do _this_, eu gosto de olhar um trecho de código e saber de pronto que o método ou campo que estou lendo/alterando não está definido na classe que estou trabalhando, o que me deixa alerta sobre uma implementação pré-existente.

**Comment by vinibeloni** [comment]
@daniellessio yep, realmente, dependendo do nome vc fica meio perdido daonde vem, mas como eu disse na maioria das vezes não muda quase nada, queria mais opinião da galera (com exemplos) pra gente chegar em uma conclusão.

**Comment by brandonluizz** [comment]
Bom, se pensarmos que teoricamente criamos classes com poucas responsabilidades e métodos que fazem apenas uma coisa, eu não vejo o porque do uso do `this`(entendo que existe os casos de necessidades e concordo com eles) para identificar de onde vem a referência. 
Se pararmos para pensar nos casos em que não utilizamos herança só podemos referenciar os métodos e propriedades da própria classe, a não ser que estamos instanciando um objeto de um outro namespace. 

Concordo com o que a galera disse sobre o uso do `this `e do `base` em casos que são necessários, mas fora isso concordo que não é preciso sua utilização.

Temos que chegar em um acordo sobre isso, se não vai morrer por aqui.

**Comment by robertjosef** [comment]
Galera, achei a discussão bem legal. 

Na minha visão, não vejo a necessidade de utilizar se não existe a obrigatoriedade. isso é apenas questão de gosto e convenção de código. 

O importante é chegar a uma conclusão e seguirmos um padrão. 


Eu tinha hábito de usar  em contrutores ou método tinha a mesma nomenclatura da classe. 

Se não me engano, o próprio compilador faz isso, mas fica oculto. 

Abração!

**Comment by daniellessio** [comment]
Pela tendência da opinião geral, não vejo problema em tirarmos a obrigatoriedade desse cara.

**Comment by LouriJr** [comment]
Tirar o obrigatoriedade ou proibir o uso?

**Comment by LouriJr** [comment]
Para mim, faz mais sentido proibir quando desnecessário. 
Pode ser que vire uma bagunça se usar em alguns contextos e em outros não. Perde-se o padrão

**Comment by vinibeloni** [comment]
quem vai levantar a briga pra usar _underline_ ? kkkkkkkkkkkkkkkkk ai fica feio demais

**Comment by LouriJr** [comment]
underscore é muito feio, namoral kkkkk

**Comment by vinibeloni** [comment]
```csharp
// com underline
private string _smurfDeBronze = "huh";

private void NaoSei()
{
  if(string.IsNullOrEmpty(_smurfDeBronze)){ throw new Expcetion("Uhu"); }
}

// sem underline
private string smurfDeBronze = "huh";

private void NaoSei()
{
  if(string.IsNullOrEmpty(smurfDeBronze)){ throw new Expcetion("Uhu"); }
}

// com this
private string smurfDeBronze = "huh";

private void NaoSei()
{
  if(string.IsNullOrEmpty(this.smurfDeBronze)){ throw new Expcetion("Uhu"); }
}
```

quero ver sangue agora
